File: app/src/main/java/igrek/todotree/ui/treelist/TreeListView.java
package igrek.todotree.ui.treelist;

import android.content.Context;
import android.graphics.Canvas;
import android.util.AttributeSet;
import android.util.SparseArray;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewTreeObserver;
import android.widget.AdapterView;
import android.widget.ListView;

import java.util.List;
import java.util.Set;

import igrek.todotree.domain.treeitem.AbstractTreeItem;
import igrek.todotree.info.logger.Logger;
import igrek.todotree.info.logger.LoggerFactory;
import igrek.todotree.intent.ItemEditorCommand;
import igrek.todotree.intent.TreeCommand;
import igrek.todotree.ui.contextmenu.ItemActionsMenu;
import igrek.todotree.ui.errorcheck.UIErrorHandler;

public class TreeListView extends ListView implements AdapterView.OnItemClickListener, AdapterView.OnItemLongClickListener {
	
	private Logger logger = LoggerFactory.INSTANCE.getLogger();
	private TreeItemAdapter adapter;
	private TreeListScrollHandler scrollHandler;
	private TreeListReorder reorder = new TreeListReorder(this);
	private TreeListGestureHandler gestureHandler = new TreeListGestureHandler(this);
	
	/** view index -> view height */
	private SparseArray<Integer> itemHeights = new SparseArray<>();
	
	public TreeListView(Context context) {
		super(context);
	}
	
	public TreeListView(Context context, AttributeSet attrs, int defStyle) {
		super(context, attrs, defStyle);
	}
	
	public TreeListView(Context context, AttributeSet attrs) {
		super(context, attrs);
	}
	
	public void init(Context context) {
		adapter = new TreeItemAdapter(context, null, this);
		scrollHandler = new TreeListScrollHandler(this, context);
		
		setOnItemClickListener(this);
		setOnItemLongClickListener(this);
		setOnScrollListener(scrollHandler);
		setChoiceMode(ListView.CHOICE_MODE_SINGLE);
		setAdapter(adapter);
	}
	
	@Override
	public TreeItemAdapter getAdapter() {
		return adapter;
	}
	
	public TreeListReorder getReorder() {
		return reorder;
	}
	
	public TreeListScrollHandler getScrollHandler() {
		return scrollHandler;
	}
	
	public TreeListGestureHandler getGestureHandler() {
		return gestureHandler;
	}
	
	@Override
	public boolean onInterceptTouchEvent(MotionEvent ev) {
		if (ev.getSource() == 777) { // from moveButton
			if (ev.getAction() == MotionEvent.ACTION_MOVE)
				return true;
		}
		return super.onInterceptTouchEvent(ev);
	}
	
	@Override
	public boolean onTouchEvent(MotionEvent event) {
		switch (event.getAction() & MotionEvent.ACTION_MASK) {
			case MotionEvent.ACTION_DOWN:
				gestureHandler.gestureStart(event.getX(), event.getY());
				break;
			case MotionEvent.ACTION_MOVE:
				if (reorder.isDragging()) {
					reorder.setLastTouchY(event.getY());
					reorder.handleItemDragging();
					return false;
				}
				break;
			case MotionEvent.ACTION_UP:
				if (gestureHandler.handleItemGesture(event.getX(), event.getY(), scrollHandler.getScrollOffset()))
					return super.onTouchEvent(event);
			case MotionEvent.ACTION_CANCEL:
				reorder.itemDraggingStopped();
				gestureHandler.reset();
				break;
		}
		return super.onTouchEvent(event);
	}
	
	public void onItemTouchDown(int position, MotionEvent event, View v) {
		gestureHandler.gestureStartPos(position);
	}
	
	@Override
	public void invalidate() {
		super.invalidate();
		if (reorder != null && reorder.isDragging()) {
			reorder.setDraggedItemView();
		}
	}
	
	@Override
	protected void dispatchDraw(Canvas canvas) {
		super.dispatchDraw(canvas);
		reorder.dispatchDraw(canvas);
	}
	
	public void setItemsAndSelected(List<AbstractTreeItem> items, Set<Integer> selectedPositions) {
		adapter.setSelections(selectedPositions);
		setItems(items);
	}
	
	private void setItems(List<AbstractTreeItem> items) {
		adapter.setDataSource(items);
		invalidate();
		calculateViewHeights();
	}
	
	public List<AbstractTreeItem> getItems() {
		return adapter.getItems();
	}
	
	private void calculateViewHeights() {
		// WARNING: for a moment - there's invalidated item heights map
		final ViewTreeObserver observer = this.getViewTreeObserver();
		observer.addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
			@Override
			public void onGlobalLayout() {
				
				itemHeights.clear();
				TreeListView.this.getViewTreeObserver().removeGlobalOnLayoutListener(this);
				// now view width should be available at last
				int viewWidth = TreeListView.this.getWidth();
				if (viewWidth == 0)
					logger.warn("List view width == 0");
				
				int measureSpecW = MeasureSpec.makeMeasureSpec(viewWidth, MeasureSpec.EXACTLY);
				for (int i = 0; i < adapter.getCount(); i++) {
					View itemView = adapter.getView(i, null, TreeListView.this);
					itemView.measure(measureSpecW, MeasureSpec.UNSPECIFIED);
					itemHeights.put(i, itemView.getMeasuredHeight());
				}
				
			}
		});
	}
	
	public int getItemHeight(int position) {
		Integer h = itemHeights.get(position);
		if (h == null) {
			logger.warn("Item View (" + position + ") = null");
			return 0;
		}
		
		return h;
	}
	
	public void putItemHeight(Integer position, Integer height) {
		itemHeights.put(position, height);
	}
	
	public View getItemView(int position) {
		return adapter.getStoredView(position);
	}
	
	@Override
	public void onItemClick(AdapterView<?> parent, final View view, int position, long id) {
		try {
			if (position == adapter.getItems().size()) {
				//nowy element
				new ItemEditorCommand().addItemClicked();
			} else {
				//istniejący element
				AbstractTreeItem item = adapter.getItem(position);
				new TreeCommand().itemClicked(position, item);
			}
		} catch (Throwable t) {
			UIErrorHandler.showError(t);
		}
	}
	
	@Override
	public boolean onItemLongClick(AdapterView<?> parent, View view, int position, long id) {
		if (!reorder.isDragging()) {
			reorder.itemDraggingStopped();
			gestureHandler.reset();
			new ItemActionsMenu(position).show(view);
		}
		return true;
	}
	
	@Override
	public int computeVerticalScrollOffset() {
		return super.computeVerticalScrollOffset();
	}
	
	@Override
	public int computeVerticalScrollExtent() {
		return super.computeVerticalScrollExtent();
	}
	
	@Override
	public int computeVerticalScrollRange() {
		return super.computeVerticalScrollRange();
	}
	
	public Integer getCurrentScrollPosition() {
		return scrollHandler.getCurrentScrollPosition();
	}
	
	public void scrollToBottom() {
		scrollHandler.scrollToBottom();
	}
	
	public void scrollToPosition(int y) {
		scrollHandler.scrollToPosition(y);
	}
	
	public void scrollToItem(int itemIndex) {
		scrollHandler.scrollToItem(itemIndex);
	}
}

File: app/src/main/java/igrek/todotree/ui/treelist/TreeListView.kt
package igrek.todotree.ui.treelist

import android.content.Context
import android.view.View
import igrek.todotree.info.logger.Logger
import igrek.todotree.ui.errorcheck.UIErrorHandler

class TreeListView : ListView, OnItemClickListener, OnItemLongClickListener {
    private val logger: Logger = LoggerFactory.logger
    private var adapter: TreeItemAdapter? = null
    var scrollHandler: TreeListScrollHandler? = null
        private set
    val reorder: TreeListReorder? = TreeListReorder(this)
    val gestureHandler = TreeListGestureHandler(this)

    /** view index -> view height  */
    private val itemHeights: SparseArray<Int> = SparseArray<Int>()

    constructor(context: Context?) : super(context) {}
    constructor(context: Context?, attrs: AttributeSet?, defStyle: Int) : super(
        context,
        attrs,
        defStyle
    ) {
    }

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {}

    fun init(context: Context?) {
        adapter = TreeItemAdapter(context!!, null, this)
        scrollHandler = TreeListScrollHandler(this, context)
        setOnItemClickListener(this)
        setOnItemLongClickListener(this)
        setOnScrollListener(scrollHandler)
        setChoiceMode(ListView.CHOICE_MODE_SINGLE)
        setAdapter(adapter)
    }

    override fun getAdapter(): TreeItemAdapter {
        return adapter!!
    }

    override fun onInterceptTouchEvent(ev: MotionEvent): Boolean {
        if (ev.getSource() == 777) { // from moveButton
            if (ev.getAction() == MotionEvent.ACTION_MOVE) return true
        }
        return super.onInterceptTouchEvent(ev)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        when (event.getAction() and MotionEvent.ACTION_MASK) {
            MotionEvent.ACTION_DOWN -> gestureHandler.gestureStart(event.getX(), event.getY())
            MotionEvent.ACTION_MOVE -> if (reorder!!.isDragging) {
                reorder.setLastTouchY(event.getY())
                reorder.handleItemDragging()
                return false
            }
            MotionEvent.ACTION_UP -> {
                if (gestureHandler.handleItemGesture(
                        event.getX(),
                        event.getY(),
                        scrollHandler!!.scrollOffset
                    )
                ) return super.onTouchEvent(event)
                reorder!!.itemDraggingStopped()
                gestureHandler.reset()
            }
            MotionEvent.ACTION_CANCEL -> {
                reorder!!.itemDraggingStopped()
                gestureHandler.reset()
            }
        }
        return super.onTouchEvent(event)
    }

    fun onItemTouchDown(position: Int, event: MotionEvent?, v: View?) {
        gestureHandler.gestureStartPos(position)
    }

    override fun invalidate() {
        super.invalidate()
        if (reorder != null && reorder.isDragging) {
            reorder.setDraggedItemView()
        }
    }

    protected override fun dispatchDraw(canvas: Canvas) {
        super.dispatchDraw(canvas)
        reorder!!.dispatchDraw(canvas)
    }

    fun setItemsAndSelected(items: List<AbstractTreeItem>, selectedPositions: Set<Int?>?) {
        adapter!!.setSelections(selectedPositions)
        setItems(items)
    }

    private fun setItems(items: List<AbstractTreeItem>) {
        adapter!!.setDataSource(items)
        invalidate()
        calculateViewHeights()
    }

    val items: List<Any>?
        get() = adapter!!.items

    private fun calculateViewHeights() {
        // WARNING: for a moment - there's invalidated item heights map
        val observer: ViewTreeObserver = this.getViewTreeObserver()
        observer.addOnGlobalLayoutListener(object : OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                itemHeights.clear()
                this@TreeListView.getViewTreeObserver().removeGlobalOnLayoutListener(this)
                // now view width should be available at last
                val viewWidth: Int = this@TreeListView.getWidth()
                if (viewWidth == 0) logger.warn("List view width == 0")
                val measureSpecW: Int = MeasureSpec.makeMeasureSpec(viewWidth, MeasureSpec.EXACTLY)
                for (i in 0 until adapter!!.count) {
                    val itemView = adapter!!.getView(i, null, this@TreeListView)
                    itemView.measure(measureSpecW, MeasureSpec.UNSPECIFIED)
                    itemHeights.put(i, itemView.measuredHeight)
                }
            }
        })
    }

    fun getItemHeight(position: Int): Int {
        val h: Int = itemHeights.get(position)
        if (h == null) {
            logger.warn("Item View ($position) = null")
            return 0
        }
        return h
    }

    fun putItemHeight(position: Int?, height: Int?) {
        itemHeights.put(position, height)
    }

    fun getItemView(position: Int): View? {
        return adapter!!.getStoredView(position)
    }

    override fun onItemClick(parent: AdapterView<*>?, view: View, position: Int, id: Long) {
        try {
            if (position == adapter!!.items!!.size) {
                //nowy element
                ItemEditorCommand().addItemClicked()
            } else {
                //istniejący element
                val item: AbstractTreeItem? = adapter!!.getItem(position)
                TreeCommand().itemClicked(position, item)
            }
        } catch (t: Throwable) {
            UIErrorHandler.showError(t)
        }
    }

    override fun onItemLongClick(
        parent: AdapterView<*>?,
        view: View,
        position: Int,
        id: Long
    ): Boolean {
        if (!reorder!!.isDragging) {
            reorder.itemDraggingStopped()
            gestureHandler.reset()
            ItemActionsMenu(position).show(view)
        }
        return true
    }

    override fun computeVerticalScrollOffset(): Int {
        return super.computeVerticalScrollOffset()
    }

    override fun computeVerticalScrollExtent(): Int {
        return super.computeVerticalScrollExtent()
    }

    override fun computeVerticalScrollRange(): Int {
        return super.computeVerticalScrollRange()
    }

    val currentScrollPosition: Int
        get() = scrollHandler!!.currentScrollPosition

    fun scrollToBottom() {
        scrollHandler!!.scrollToBottom()
    }

    fun scrollToPosition(y: Int) {
        scrollHandler!!.scrollToPosition(y)
    }

    fun scrollToItem(itemIndex: Int) {
        scrollHandler!!.scrollToItem(itemIndex)
    }
}